### PR TITLE
Fix issue with docker-pull and authentication 

### LINKED ISSUE #4055: Kaniko: docker-pull fails to authenticate to container registry to pull container artifact
**Describe the bug**

Attempting to use the `docker-pull` plugin with on-demand runners, which uses Kaniko to pull containers, ends up failing to authenticate to the container registry despite being configured with a username and password via Waypoint:

```hcl
app "tetris" {
  build {
    use "docker" {
    }
    workspace "production" {
      use "docker-pull" {
        image = var.image
        tag   = var.tag
        encoded_auth = base64encode(
          jsonencode({
            username = var.registry_username,
            password = var.registry_password
          })
        )
      }
    }
    registry {
      use "docker" {
        image    = var.image
        tag      = var.tag
        username = var.registry_username
        password = var.registry_password
        local    = var.registry_local
      }
    }
  }
...
...
```

Note that `var.registry_username` and `var.registry_password` are the same input variables used to _push_ the container to that same registry, so I know at least internally those vars are set.

```
✓ Running build v3
✓ Injecting entrypoint...
⠹ Executing Kaniko...
 │ INFO[0000] Retrieving image manifest <registry>/tetris:latest
 │ INFO[0000] Retrieving image <registry>/tetris:latest from registry <registry>
 │ error building image: GET https://<registry>/token?scope=repository%3Atetris%3Ap
 │ ull&service=<registry>: : Authentication is required
! 1 error occurred:
        * rpc error: code = Unknown desc = exit status 1

```

**Steps to Reproduce**

Attempt to run a docker-pull plugin for a build step using on-demand runners via Kaniko.

**Expected behavior**

Kaniko should have the proper credentials setup to pull a container from the specified registry.

**Waypoint Platform Versions**
Additional version and platform information to help triage the issue if
applicable:

* Waypoint CLI Version: 0.10.2
* Waypoint Server Platform and Version: n/a
* Waypoint Plugin: `docker-pull` with `kaniko`

**Additional context**

There are some closely related issues but they don't 100% line up with this one:

- https://github.com/hashicorp/waypoint/issues/4001
- https://github.com/hashicorp/waypoint/issues/3557

### PR DESCRIPTION
**Pull request by catsby**
Fix an issue where both `docker-pull` and subsequent `registry` blocks needed separate authentication. 

Example:

```hcl
app "some-app" {
  build {
    use "docker" {
      // Use the included Dockerfile
    }
    workspace "production" {
      use "docker-pull" {
        image = "some-private-registry/image"
        tag   = "stable"
        encoded_auth = base64encode(
          jsonencode({
            username = "some-username",
            password = "some-password"
          })
        )
      }
    }

    registry {
      use "docker" {
        image    = "somerepo/image"
        tag      = "latest"
        username = "some-username"
        password = "some-password"
      }
    }
  }
  
  [...]
}
```

Prior to these changes, Waypoint could only authenticate with the registry in the `registry` block, and did not properly authenticate with the registry that the `docker-pull` block would eventually be talking to. 

In this PR we create a 2nd [OCI Proxy Server](https://github.com/hashicorp/waypoint/blob/b0017806980a11f93d9c2411815818a123ae6ccd/internal/pkg/epinject/ociregistry/server.go#L39) to handle authentication with the registry we pull from, and modify the resulting generate `Dockerfile` to then pull from the proxy and not directly from the source. 

Along the way we ran into some issues with manifests so we had to make some changes to the proxy server itself as well. 

Fixes https://github.com/hashicorp/waypoint/issues/4055

### CODE REVIEW

File: builtin/docker/pull/kaniko.go [hunk 1]
```diff
@@ -41,98 +41,177 @@ func (b *Builder) pullWithKaniko(
...
+	{
+		if b.config.EncodedAuth != "" {
+			//If EncodedAuth is set, use that
+			user, pass, err := wpdocker.CredentialsFromConfig(b.config.EncodedAuth)
+			if err != nil {
+				return nil, err
+			}
+			sourceProxy.AuthConfig.Username = user
+			sourceProxy.AuthConfig.Password = pass
+		} else if *b.config.Auth != (wpdocker.Auth{}) {
```
Comment (evanphx) [inline builtin/docker/pull/kaniko.go:163]: If `Auth` is nil, this will panic I believe.

File: internal/pkg/epinject/ociregistry/server.go [hunk 2]
```diff
@@ -265,19 +265,8 @@ func (s *Server) FetchBlob(name string, id string) ([]byte, error) {
...
-	found, err := s.probeBlob(s.entrypointRepo, s.entrypointId)
-	if err != nil {
-		return err
-	}
-
-	if found {
-		s.Logger.Debug("entrypoint layer already exists, not reuploading")
-		return nil
-	}
-
```
Comment (evanphx) [inline internal/pkg/epinject/ociregistry/server.go:277]: We can put this back in, it will make things faster and wasn't causing an issue.

**Review by evanphx — CHANGES_REQUESTED** [review]
A couple of minor things.

**Review by briancain — APPROVED** [review]
I agree with Evans feedback, but wanted to approve because I gave this a test and it worked and solves the bug! ✨ 🐛 🐳

File: builtin/docker/pull/kaniko.go (hunk 1 above)
Comment (catsby) [reply builtin/docker/pull/kaniko.go]: Added a nil check in e5b5fe97c

File: internal/pkg/epinject/ociregistry/server.go (hunk 2 above)
Comment (catsby) [reply internal/pkg/epinject/ociregistry/server.go]: Restored in 4480ed51f

**Comment by catsby** [comment]
I addressed the feedback, @evanphx please take another look when you can, thanks!

**Comment by catsby** [comment]
@evanphx please take another look when you can

**Review by evanphx — APPROVED** [review]

**Pull request merged**